code/edgeidentity/src/main/java/com/adobe/marketing/mobile/edge/identity/IdentityExtension.java

```diff
@@ -13,232 +13,206 @@
 
 import static com.adobe.marketing.mobile.edge.identity.IdentityConstants.LOG_TAG;
 
+import androidx.annotation.NonNull;
+import androidx.annotation.VisibleForTesting;
 import com.adobe.marketing.mobile.Event;
 import com.adobe.marketing.mobile.Extension;
 import com.adobe.marketing.mobile.ExtensionApi;
-import com.adobe.marketing.mobile.ExtensionError;
-import com.adobe.marketing.mobile.ExtensionErrorCallback;
 import com.adobe.marketing.mobile.LoggingMode;
 import com.adobe.marketing.mobile.MobileCore;
+import com.adobe.marketing.mobile.SharedStateResolution;
+import com.adobe.marketing.mobile.SharedStateResult;
+import com.adobe.marketing.mobile.util.StringUtils;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.concurrent.ConcurrentLinkedQueue;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
 
 class IdentityExtension extends Extension {
 
-	private ExecutorService executorService;
-	private final Object executorMutex = new Object();
-	private final ConcurrentLinkedQueue<Event> cachedEvents; // cached events in memory until required shared states are resolved
+	/**
+	 * A {@code SharedStateCallback} to retrieve the last set state of an extension and to
+	 * create an XDM state at the event provided.
+	 */
+	private final SharedStateCallback sharedStateHandle = new SharedStateCallback() {
+		@Override
+		public SharedStateResult getSharedState(final String stateOwner, final Event event) {
+			return getApi().getSharedState(stateOwner, event, false, SharedStateResolution.LAST_SET);
```

Comment:
Using `LAST_SET` here is a change from the previous implementation which would essentially get `ALL` states. I think this change is okay, however, but I did just wanted to point out the change in case it wasn't intentional. In most cases, the sharedStateHandle is being used for bootup here and in IdentityState, so using `LAST_SET` is okay. The one case where we'd want to ensure we get the last state is `handleIdentityDirectECIDUpdate`. However, there, it's pretty much guaranteed the shared state is set before the shared state update event is dispatched, so using `LAST_SET` there is fine too.

Reply:
Your explanation is spot on! This change was intentional and was made to accommodate the current event dispatch implementation in core where the extensions can process events at their own speed as well as the need in EdgeIdentity usages to get a shared state only if it was set.